### ISSUE TITLE
support using KSM as fee(cases: transfer back asset to Statemine)

### ISSUE THREAD

**Issue by zqhxuyuan** [open-web3-stack/open-runtime-module-library#698]
current, we have multi assets support, and use `fee_item` asset as reserve

https://github.com/open-web3-stack/open-runtime-module-library/blob/4a66b299037cc3997689538f82847785f9afa65d/xtokens/src/lib.rs#L590

i.e. if `currencies=vec![(CurrencyId::R, 100), (CurrencyId::B, 450)]`, and fee_item=0, reserve=Parent, because fee asset is R, then xcm sent to Relaychain likes:
```
origin: MultiLocation { parents: 0, interior: X1(Parachain(1)) }, 
message: Xcm([
    WithdrawAsset(MultiAssets([
        MultiAsset { id: Concrete(MultiLocation { parents: 0, interior: Here }), fun: Fungible(100) }, 
        MultiAsset { id: Concrete(MultiLocation { parents: 0, interior: X2(Parachain(2), GeneralKey([66])) }), fun: Fungible(450) }])), // 🔥
    ClearOrigin, 
    BuyExecution { fees: MultiAsset { id: Concrete(MultiLocation { parents: 0, interior: Here }), fun: Fungible(50) }, weight_limit: Limited(40) }, 
    DepositReserveAsset { 
        assets: Wild(All), max_assets: 2, 
        dest: MultiLocation { parents: 0, interior: X1(Parachain(2)) }, 
        xcm: Xcm([
            BuyExecution { fees: MultiAsset { id: Concrete(MultiLocation { parents: 1, interior: Here }), fun: Fungible(50) }, weight_limit: Limited(40) }, 
            DepositAsset { assets: Wild(All), max_assets: 2, beneficiary: MultiLocation { parents: 0, interior: X1(AccountId32 { network: Any, id: [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1] }) } }]) }
]), weight_limit: 18446744073709551615, weight_credit: 0
```
this will error when execute second WithdrawAsset because relaychain doesn't have `(0, Parachain(2), GeneralKey(B))` asset.

---

if currencies=`vec![(CurrencyId::R, 100), (CurrencyId::B, 450)]`, and fee_item=1, reserve=`(1,Parachain(2))`, because fee asset is B, then xcm send to Sibling Parachain(2) likes:
```
origin: MultiLocation { parents: 0, interior: X1(Parachain(1)) }, 
message: Xcm([
    WithdrawAsset(MultiAssets([
        MultiAsset { id: Concrete(MultiLocation { parents: 0, interior: X1(GeneralKey([66])) }), fun: Fungible(450) }, // sibling_a_account should have enougn B token
        MultiAsset { id: Concrete(MultiLocation { parents: 1, interior: Here }), fun: Fungible(100) }])), // sibling_a_account also should have enough R token
    ClearOrigin, 
    BuyExecution { 
        fees: MultiAsset { 
            id: Concrete(MultiLocation { parents: 0, interior: X1(GeneralKey([66])) }), fun: Fungible(450)  // ⬅️
        }, weight_limit: Limited(40) }, 
    DepositAsset { 
        assets: Wild(All), max_assets: 2, 
        beneficiary: MultiLocation { parents: 0, interior: X1(AccountId32 { network: Any, id: [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1] }) } 
	}
]), weight_limit: 6148914691236517205, weight_credit: 0
```
this can sucess if we open gate here(if not, there'll be `DistinctReserveForAssetAndFee` error): 

https://github.com/open-web3-stack/open-runtime-module-library/blob/4a66b299037cc3997689538f82847785f9afa65d/xtokens/src/lib.rs#L584-L587

and the xcm above use `(0, GeneralKey(B))` as fee, although this is what the fee_item exactly wanted.

---

**but in some cases we want use sibling parachain as reserve chain, and also use relaychain asset as fee.** 
i.e. transfer RMRK from Karura to Statemine, we use KSM as fee. the xcm on recipient parachain should be like:
```
message: Xcm([
    WithdrawAsset(MultiAssets([
        MultiAsset { id: Concrete(MultiLocation { parents: 0, interior: X1(GeneralKey([66])) }), fun: Fungible(450) }, // sibling_a_account should have enougn B token
        MultiAsset { id: Concrete(MultiLocation { parents: 1, interior: Here }), fun: Fungible(100) }])), // sibling_a_account also should have enough R token
    ClearOrigin, 
    BuyExecution { 
        fees: MultiAsset { 
            id: Concrete(MultiLocation { parents: 1, interior: Here }), fun: Fungible(50)  // ⬅️
        }, weight_limit: Limited(40) }, 
    DepositAsset { 
        assets: Wild(All), max_assets: 2, 
        beneficiary: MultiLocation { parents: 0, interior: X1(AccountId32 { network: Any, id: [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1] }) } 
	}
])
```

---

PS: in the case of transfer RMRK from Karura to Statemine, if we transfer only RMRK, in the Statemine side, as trader not matched, xcm executed in Statemine will failed, and throw TooExpsensive error. so we must use KSM as fee, because Statemine only support KSM as fee for now.

https://github.com/paritytech/cumulus/blob/86f76c5619c64d1300315612695ad4b4fcd0f562/polkadot-parachains/statemine/src/lib.rs#L557

**Comment by zqhxuyuan** [comment]
@girazoki @xlc I think previous we've some discussion in this issue: https://github.com/open-web3-stack/open-runtime-module-library/issues/663

for this PR I think we could allow fund KSM to parachain sovereign account on Statemine. but of course the weakness is If someone in Karura(sender parachain) set too much KSM as fee, 
the parachain sovereign account will soon be exhausted so that other user can't transfer RMRK back to Statemine. 

but consider there're issue ongoing: https://github.com/paritytech/cumulus/issues/892 which allow Statemine transfer RMRK and use KRM as fee to Karura.
and if that issue is fixed, that'll be problem too as you mentioned in case one(allow transfer KSM from Statemine). this need discuss more sufficient or audit.

**Comment by girazoki** [comment]
Just a heads up that https://github.com/paritytech/polkadot/pull/4756/files might solve our issues. If the transactor receives an origin parameter, we can simply mint a different token if the transfer came from Statemine, which would simplify all the issues we discussed above

**Comment by zqhxuyuan** [comment]
> if you have the polkadotXCM pallet enabled I think a user could craft the request
even though we can disable polkadotXCM by call filter as Acala current did, but as xTokens is still exposed with the new PR, I'm afraid we can't stop user/bot setting too much KSM fee when transfer asset back to Statemine. 

unless we set fixed KSM fee(which relay on [trader config on Statemine](https://github.com/paritytech/cumulus/blob/e6a2cfa692aa002881b7fd2b645b8095cd68e37d/polkadot-parachains/statemine/src/xcm_config.rs#L164-L165)) in the code. I guess that's why bifrost @yrong previous did: https://github.com/bifrost-finance/bifrost/pull/420 instead of specify fee amount to prevent above potential attack happend.

**Comment by yrong** [comment]
Yes, remeber some discussion with @girazoki before in https://github.com/bifrost-finance/bifrost/pull/420, ksm from statemine should not be mapped to the same token as from kusama. Our workround now is to deduct extra fee and then manually fund our sovereign account in statemine as compensation. I do think there should be some better solution.

**Comment by girazoki** [comment]
I see, thanks for the clarification guys! that makes total sense. 

I think there will be a way after xcm v3, since assetTransactors will receive the context of the call,which includes the origin. Therefore we can map the parent token to a different token depending if we receive it from Statemine or the Relay.

**Comment by zqhxuyuan** [comment]
After talk to @shaunxw and @xlc, we got two solution for this case.
1. one xcm direct send to Statemine, which need pre-fund relaychain asset to parachain account on Statemine, also let all unused fee to parachain account instead of recipient. but this still has the problem that parachain account will exhausted and need fund again.
2. two xcm: first one send to relaychain and route to Statemine(A-[R]-B NonReserve case), transfer user's fee to parachain account on Statemine. second one like first solution(A-[R,B]-B, ToReserve case). but in this second solution, we don't need to pre-fund relaychain asset to parachain account on Statemine.

we choose second solution for now(detail see the PR), the bad thing for user is need more fee, the good thing for parachain is they would't need pre-fund anymore.

**Comment by xlc** [comment]
closed by #700
